<!DOCTYPE html>
<html lang="" xml:lang="">
  <head>
    <title>Scatter/gather-operations in Snakemake</title>
    <meta charset="utf-8" />
    <script src="scatter-gather_files/header-attrs-2.8/header-attrs.js"></script>
    <link href="scatter-gather_files/remark-css-0.0.1/default.css" rel="stylesheet" />
    <link rel="stylesheet" href="/Users/john/git/nbis/workshop-snakemake-byoc/lectures/template.css" type="text/css" />
  </head>
  <body>
    <textarea id="source">


layout: true
&lt;div class="scilife-logo"&gt;&lt;/div&gt;
&lt;div class="nbis-logo"&gt;&lt;/div&gt;

---

class: center, middle

.HUGE[Scatter/gather-operations]
&lt;br&gt;
.HUGE[in Snakemake]



---

# What does scatter/gather mean?

--

* .green[Scatter]: turn input into several pieces of output

--

* .green[Gather]: bring together (aggregate) results from the different pieces


--

Snakemake now has built-in support for scatter/gather processes via the `scattergather` directive. Described further in the documentation: [Defining scatter-gather processes](https://snakemake.readthedocs.io/en/stable/snakefiles/rules.html#defining-scatter-gather-processes). Currently not very flexible though.

---

# When are scatter-gather processes handy?

--

- demultiplexing sequencing runs

  - multiple samples per plate
  - split plates into separate files per sample

--

- extract reads from bam files

  - reads mapped to several genomes
  - split sequences per genome

--

- parallelize analyses

  - _e.g._ multiple sequences per sample
  - split input into smaller chunks and run analyses in parallell

_etc_...

--

Between scattering and gathering there's some type of analyses performed.

---

# The basics

```python
DATASETS = ["a", "b", "c"]

rule scatter:
    output:
        expand('{dataset}.txt', dataset=DATASETS)
    input:
        data = 'data.tar.gz'
    shell:
        """
        tar xvf {input}
        """

rule uppercase:
    input:
        "{dataset}.txt"
    output:
        "{dataset}.uppercase.txt"
    shell:
        """
        tr [a-z] [A-Z] &lt; {input} &gt; {output}
        """

rule gather:
    output:
        "aggregated.txt"
    input:
        expand("{dataset}.uppercase.txt", dataset=DATASETS)
    shell:
        """
        cat {input} &gt; {output}
        """
```
--

```bash
snakemake -c 1 

Job stats:
job          count    min threads    max threads
---------  -------  -------------  -------------
gather           1              1              1
scatter          1              1              1
uppercase        3              1              1
total            5              1              1
```

---

# The basics

.center[&lt;img src="filegraph.png" width=20%/&gt;]

---

# Example: split files for parallelization

--
- one fastq file per sample
```
data
├── sample1.fastq
└── sample2.fastq
```
--
- split into several files (scatter)
```
splits
├── sample1
│   ├── sample1.1.fastq
│   ├── sample1.2.fastq
│   ├── sample1.3.fastq
|   ├── sample1.4.fastq
│   └── sample1.5.fastq
├── sample2
|   ├── sample2.1.fastq
|   ├── sample2.2.fastq
|   ├── sample2.3.fastq
|   ├── sample2.4.fastq
└   └── sample2.5.fastq
```
---
# Example: split files for parallelization

- process individual files (parallelization)
```
rc
├── sample1
│   ├── sample1.1.rc.fastq
│   ├── sample1.2.rc.fastq
│   ├── sample1.3.rc.fastq
|   ├── sample1.4.rc.fastq
│   └── sample1.5.rc.fastq
├── sample2
|   ├── sample2.1.rc.fastq
|   ├── sample2.2.rc.fastq
|   ├── sample2.3.rc.fastq
|   ├── sample2.4.rc.fastq
└   └── sample2.5.rc.fastq
```
--
- aggregate results (gather)
```
sample1.rc.fastq
sample2.rc.fastq
```

---
# Example: split files for parallelization

We start with defining the number of splits
```python
splits = 5
scatteritems = range(1, splits+1]
```

--
Then define a rule to scatter each sample fastq
```python
rule scatter:
    output:
        expand("splits/{{sample}}/{{sample}}.{scatteritem}.fastq", scatteritem = scatteritems)
    input:
        "data/{sample}.fastq"
    log:
        "logs/{sample}.scatter.log"
    conda:
        "envs/seqkit.yml"
    params:
        parts = splits,
        outdir = lambda wildcards, output: os.path.dirname(output[0])
    shell:
        """ 
        seqkit split -p {params.parts} -O {params.outdir} {input} &gt; {log} 2&gt;&amp;1
        rename 's/part_0*//' {params.outdir}/{wildcards.sample}.*.fastq
        """
```

Here `scatteritem` is not a wildcard because it is expanded using the `scatteritems` list.

---
# Example: split files for parallelization

Next, a rule to do something with the split files per sample

```python
rule reversecomplement:
    output:
        "rc/{sample}/{sample}.{scatteritem}.rc.fastq"
    input:
        "splits/{sample}/{sample}.{scatteritem}.fastq"
    conda:
        "envs/seqkit.yml"
    shell:
        """
        seqkit seq --reverse --complement {input} &gt; {output}
        """
```

Here both `scatteritem` and `sample` are wildcards. The rule is generalized to work on any value for these wildcards.

---
# Example: split files for parallelization

Then a rule to gather the results per sample

```python
rule gather:
    output:
        "{sample}.rc.fastq"
    input:
        expand("rc/{{sample}}/{{sample}}.{scatteritem}.rc.fastq", scatteritem = scatteritems)
    shell:
        "cat {input} &gt; {output}"
```

Here `scatteritem` is not a wildcard, but `sample` is. The rule can gather split files for any sample.

---
# Example: split files for parallelization

Finally we put everything together, and define a pseudo rule 'all' that takes as input the gathered results for
all samples.

```python
samples = ["sample1", "sample2"]

splits = 5
scatteritems = range(1, splits+1)

rule all:
    input:
        expand("{sample}.rc.fastq", sample = samples)

rule scatter:
    output:
        expand("splits/{{sample}}/{{sample}}.{scatteritem}.fastq", scatteritem = scatteritems)
    input:
        "data/{sample}.fastq"

rule reversecomplement:
    output:
        "rc/{sample}/{sample}.{scatteritem}.rc.fastq"
    input:
        "splits/{sample}/{sample}.{scatteritem}.fastq"

rule gather:
    output:
        "{sample}.rc.fastq"
    input:
        expand("rc/{{sample}}/{{sample}}.{scatteritem}.rc.fastq", scatteritem = scatteritems)
```
---
# Example: split files for parallelization

```bash
snakemake -c 1 --use-conda

Building DAG of jobs...
Job stats:
job                  count    min threads    max threads
-----------------  -------  -------------  -------------
all                      1              1              1
gather                   2              1              1
reversecomplement       10              1              1
scatter                  2              1              1
total                   15              1              1
```
--

.center[&lt;img src="dag_scatter.png" width=100%/&gt;]

--

This example workflow is available at the course GitHub repository: [workshop-snakemake-byoc/tree/main/lectures/scatter-gather/](https://github.com/NBISweden/workshop-snakemake-byoc/tree/master/lectures/example-workflow)

---


class: center, middle

.HUGE[Questions?]
    </textarea>
<style data-target="print-only">@media screen {.remark-slide-container{display:block;}.remark-slide-scaler{box-shadow:none;}}</style>
<script src="https://remarkjs.com/downloads/remark-latest.min.js"></script>
<script>var slideshow = remark.create({
"slideNumberFormat": ""
});
if (window.HTMLWidgets) slideshow.on('afterShowSlide', function (slide) {
  window.dispatchEvent(new Event('resize'));
});
(function(d) {
  var s = d.createElement("style"), r = d.querySelector(".remark-slide-scaler");
  if (!r) return;
  s.type = "text/css"; s.innerHTML = "@page {size: " + r.style.width + " " + r.style.height +"; }";
  d.head.appendChild(s);
})(document);

(function(d) {
  var el = d.getElementsByClassName("remark-slides-area");
  if (!el) return;
  var slide, slides = slideshow.getSlides(), els = el[0].children;
  for (var i = 1; i < slides.length; i++) {
    slide = slides[i];
    if (slide.properties.continued === "true" || slide.properties.count === "false") {
      els[i - 1].className += ' has-continuation';
    }
  }
  var s = d.createElement("style");
  s.type = "text/css"; s.innerHTML = "@media print { .has-continuation { display: none; } }";
  d.head.appendChild(s);
})(document);
// delete the temporary CSS (for displaying all slides initially) when the user
// starts to view slides
(function() {
  var deleted = false;
  slideshow.on('beforeShowSlide', function(slide) {
    if (deleted) return;
    var sheets = document.styleSheets, node;
    for (var i = 0; i < sheets.length; i++) {
      node = sheets[i].ownerNode;
      if (node.dataset["target"] !== "print-only") continue;
      node.parentNode.removeChild(node);
    }
    deleted = true;
  });
})();
(function() {
  "use strict"
  // Replace <script> tags in slides area to make them executable
  var scripts = document.querySelectorAll(
    '.remark-slides-area .remark-slide-container script'
  );
  if (!scripts.length) return;
  for (var i = 0; i < scripts.length; i++) {
    var s = document.createElement('script');
    var code = document.createTextNode(scripts[i].textContent);
    s.appendChild(code);
    var scriptAttrs = scripts[i].attributes;
    for (var j = 0; j < scriptAttrs.length; j++) {
      s.setAttribute(scriptAttrs[j].name, scriptAttrs[j].value);
    }
    scripts[i].parentElement.replaceChild(s, scripts[i]);
  }
})();
(function() {
  var links = document.getElementsByTagName('a');
  for (var i = 0; i < links.length; i++) {
    if (/^(https?:)?\/\//.test(links[i].getAttribute('href'))) {
      links[i].target = '_blank';
    }
  }
})();</script>

<script>
slideshow._releaseMath = function(el) {
  var i, text, code, codes = el.getElementsByTagName('code');
  for (i = 0; i < codes.length;) {
    code = codes[i];
    if (code.parentNode.tagName !== 'PRE' && code.childElementCount === 0) {
      text = code.textContent;
      if (/^\\\((.|\s)+\\\)$/.test(text) || /^\\\[(.|\s)+\\\]$/.test(text) ||
          /^\$\$(.|\s)+\$\$$/.test(text) ||
          /^\\begin\{([^}]+)\}(.|\s)+\\end\{[^}]+\}$/.test(text)) {
        code.outerHTML = code.innerHTML;  // remove <code></code>
        continue;
      }
    }
    i++;
  }
};
slideshow._releaseMath(document);
</script>
<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
(function () {
  var script = document.createElement('script');
  script.type = 'text/javascript';
  script.src  = 'https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML';
  if (location.protocol !== 'file:' && /^https?:/.test(script.src))
    script.src  = script.src.replace(/^https?:/, '');
  document.getElementsByTagName('head')[0].appendChild(script);
})();
</script>
  </body>
</html>
